collector/ntp.go

```diff
@@ -17,62 +17,145 @@ package collector
+	ntpIPTTL           = kingpin.Flag("collector.ntp.ip-ttl", "IP TTL to use while sending NTP query").Default("1").Int()
+	// 3.46608s ~ 1.5s + PHI * (1 << maxPoll), where 1.5s is MAXDIST from ntp.org, it is 1.0 in RFC5905
+	// max-distance option is used as-is without phi*(1<<poll)
+	ntpMaxDistance     = kingpin.Flag("collector.ntp.max-distance", "Max accumulated distance to the root").Default("3.46608s").Duration()
+	ntpOffsetTolerance = kingpin.Flag("collector.ntp.local-offset-tolerance", "Offset between local clock and local ntpd time to tolerate").Default("1ms").Duration()
+
+	leapMidnight time.Time
 )
 
 type ntpCollector struct {
-	drift, stratum typedDesc
+	stratum, leap, rtt, offset, reftime, rootDelay, rootDispersion, sanity typedDesc
 }
 
 func init() {
 	Factories["ntp"] = NewNtpCollector
 }
 
-// NewNtpCollector returns a new Collector exposing the offset between ntp and
-// the current system time.
+// NewNtpCollector returns a new Collector exposing sanity of local NTP server.
+// Default definition of "local" is:
+// - collector.ntp.server address is a loopback address (or collector.ntp.server-is-mine flag is turned on)
+// - the server is reachable with outgoin IP_TTL = 1
 func NewNtpCollector() (Collector, error) {
-	warnDeprecated("ntp")
-	if *ntpServer == "" {
-		return nil, fmt.Errorf("no NTP server specified, see -collector.ntp.server")
+	ipaddr := net.ParseIP(*ntpServer)
+	if !*ntpServerIsLocal && (ipaddr == nil || !ipaddr.IsLoopback()) {
+		return nil, fmt.Errorf("only IP address of local NTP server is valid for -collector.ntp.server")
 	}
+
 	if *ntpProtocolVersion < 2 || *ntpProtocolVersion > 4 {
 		return nil, fmt.Errorf("invalid NTP protocol version %d; must be 2, 3, or 4", *ntpProtocolVersion)
 	}
 
+	if *ntpOffsetTolerance < 0 {
+		return nil, fmt.Errorf("Offset tolerance must be non-negative")
+	}
+
 	return &ntpCollector{
-		drift: typedDesc{prometheus.NewDesc(
-			prometheus.BuildFQName(Namespace, "ntp", "drift_seconds"),
-			"Time between system time and ntp time.",
-			nil, nil,
-		), prometheus.GaugeValue},
 		stratum: typedDesc{prometheus.NewDesc(
 			prometheus.BuildFQName(Namespace, "ntp", "stratum"),
-			"NTP server stratum.",
+			"NTPD stratum.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		leap: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "leap"),
+			"NTPD leap second indicator, 2 bits.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		rtt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "rtt_seconds"),
+			"RTT to NTPD.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		offset: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "offset_seconds"),
+			"ClockOffset between NTP and local clock.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		reftime: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "ntp", "reftime"),
```

Comment:
This should have _seconds suffix too. Maybe call it reference_timestamp_seconds? All our timestamps are usually called xx_timestamp_seconds.

Reply:
> xx_timestamp_seconds

Okay. I just followed naming convention used by series I've seen
- stat_linux with `node_boot_time`
- time with `node_time`
There are also some series I've not seen
- exec_bsd with `...boot_timestamp_seconds`
- runint with `...state_last_change_timestamp_seconds`

I don't feel like there is consistency here, but I'm missing something for sure :) Also, it's a bit confusing that node_exporter exports `node_boot_time` and `...boot_timestamp_seconds` depending on kernel, but that's probably due to backward compatibility...

<troll-mode>
Should I rename `node_time` to `node_time_timestamp_seconds` as well? :-D
</troll-mode>